Add Image Content Block

milljoniaer opened the pull request:
### What is included
ImageContentBlock component with description at the right, left or bottom.
There can also be a Creator Tag at the bottom right.

On mobile devices there will always the bottom Configuration be displayed.

---

### Screenshots
![image](https://user-images.githubusercontent.com/78978542/143577894-178568f7-6397-4029-ad57-1f2f4d4f3098.png)

![image](https://user-images.githubusercontent.com/78978542/143577962-bf553713-5fa3-4ae3-b09b-f1f5aab6b0c3.png)

![image](https://user-images.githubusercontent.com/78978542/143578029-342f1d26-d437-4e43-bae5-433d4e682fbe.png)

milljoniaer committed:
chore: set up new files for component

milljoniaer committed:
Merge branch 'main' into feat/image-contentblock

milljoniaer committed:
feat: add ImageBlockContent

milljoniaer committed:
Merge branch 'main' into feat/image-contentblock

ghost commented:
# [Gatsby Cloud](https://www.gatsbyjs.com) Build Report
**satellytes.com-main**

:tada: Your build was successful! [See the Deploy preview here.](https://satellytescommain21751-featimagecontentblock.gtsb.io)

## Build Details

[View the build logs here.](https://gatsbyjs.com/dashboard/f215130d-15e3-4176-9252-e13166fe48de/sites/674be368-adf0-4ac5-8fae-f9a4493adc51/builds/3bf52f38-8dd1-4f98-af99-3a88e864a7ee/details)

:clock1: Build time: **1m**

## Performance

### Lighthouse report

| Metric | Score |
| ----- | ---- |
| Performance |   :large_orange_diamond: 79   |
| Accessibility |   :green_heart: 100   |
| Best Practices |   :green_heart: 100   |
| SEO |   :large_orange_diamond: 85   |

:link: [View full report](https://build-3bf52f38-8dd1-4f98-af99-3a88e864a7ee.gtsb.io/reports/lighthouse/index.html)

milljoniaer committed:
chre: radio types

feedm3 commented on src/@types/global.d.ts:
```diff
@@ -3,6 +3,7 @@ declare module '*.woff';
 declare module '*.eot';
 declare module '*.ttf';
 declare module '*.png';
+declare module '.*.jpg';
```
Please use the same pattern as for other file types: `*.jpg`

milljoniaer replied:
Oops, typo :+1:

feedm3 commented on src/components/image-contentBlock/image-contentBlock.stories.tsx:
```diff
@@ -0,0 +1,37 @@
+import React from 'react';
```
Only use kebap-case filenames, don't mix with camel-case: `image-content-block.*`

milljoniaer replied:
I am sorry, I have forgotten it another time...
Fixed it, and will remember it

feedm3 commented on src/components/image-contentBlock/image-contentBlock.tsx:
```diff
@@ -0,0 +1,116 @@
+    textAlign === 'bottom' ? 'width 0; min-width:100%;' : ''}
+  ${({ textAlign }) => (textAlign !== 'bottom' ? 'max-width: 228px;' : '')}
+
+  ${down('sm')} {
```
Default CSS is always mobile (mobile first principle). So instead of using `down('sm')`, we use `up(md)` for desktop styles. 

There are some exceptions when you can use desktop first, but it shouldn't be necessary here.

milljoniaer replied:
Okay, changed it

feedm3 reviewed (dismissed):
Good first version.

Usually, we would not use the native `img` tag but always use the `GatsbyImage` component. Not sure if it's possible, but can you add a Storybook example with `GatsybImage`? If not, is it already possible to use this component on the old [office page](https://satellytes.com/office/)?

georgiee commented on src/components/image-contentBlock/image-contentBlock.stories.tsx:
```diff
@@ -0,0 +1,37 @@
+
+export default {
+  component: ImageContentBlock,
+  title: 'Components/ImageContentBlock',
```
That name is no good. Check the other names, do you see a pattern? 

![Screen Shot 2021-11-29 at 06 43 44](https://user-images.githubusercontent.com/1701755/143814990-52ddb36d-295c-411d-92a4-2221262d7db1.png)


We always remove the `ContentBlock` postfix. It's of no use for use. Please rename including the name of the component. I think `<Image/>` will work for us as it differs from `GatsbyImage` if not then we can still rename in the future.

I now see another pattern, you named `HeaderBlock` with a postfix `Block` as it would collide with the actual `Header` probably? I guess you just followed the idea for naming here? I would go backwards and say, that we should rename that component before using it extensively though. Something like `ArticleHeader` as it's the text at the beginning of an article in the best case. The designer took a shortcut and used this component then for entire content blocks (like on Services) which blurs the borders too much. Anyway, I have created an issue for that: https://github.com/satellytes/satellytes.com/issues/220

georgiee commented on src/components/image-contentBlock/image-contentBlock.tsx:
```diff
@@ -0,0 +1,116 @@
+      ? 'row'
+      : textAlign === 'left'
+      ? 'row-reverse'
+      : 'column'};
```
this is not readable at all, please create a local function at the top of the file to "calculate" instead of having such an inline definition. I don't like a single ternary operator for readability reasons, chaining it makes it worse so I suggest to use a dedicated `if` or surch in the function.

milljoniaer replied:
Thanks for the tip, I implemented it :+1:

georgiee commented on src/components/image-contentBlock/image-contentBlock.tsx:
```diff
@@ -0,0 +1,116 @@
+import { TextStyles } from '../typography/typography-v2';
+
+const BlockWrapper = styled.div<{
+  textAlign: 'right' | 'left' | 'bottom';
```
I think we don't have any rules here yet, but I prefer an isolated `BlockWrapperProps` instead of inlining the definition. You can that type (use an interface though) just before this line.

milljoniaer replied:
I actually thought about that, I have tried the "interface-way" in another component, and was thinking if I like the way with less variables better, but I have implemented it using an interface

georgiee replied:
Less is not better here as it makes reading the code more difficult because you mix two very different things. Usually we have defined rules that ideally are enforced through the linter for things like that, so your machine would complain. Right now we don't have any rule, so let's see how you think about that the next time.

milljoniaer replied:
👍

georgiee commented on src/components/image-contentBlock/image-contentBlock.tsx:
```diff
@@ -0,0 +1,116 @@
+    ${TextStyles.textXS}
+
+    margin: 12px 0;
+    width 0;
```
I think that's not good here

milljoniaer replied:
Oops, thanks, fixed it

georgiee commented on src/components/image-contentBlock/image-contentBlock.tsx:
```diff
@@ -0,0 +1,116 @@
+  }
+
+  ${theme.palette.text.default};
+`;
```
I think you arrived at a place that works by hard work and trying out a lot. I would ask you to refactor the component by using either flex or grid. This will make it much more elegant. In both worlds you can use an attribute `order` to change the order. If you don't like any layout (like in the text below image) just disable the flex or grid layout.

georgiee replied:
@milljoniaer any thoughts on my feedback?

milljoniaer replied:
I am using flex, but afaik I did that also before, but I now used gap for the space between the elements. The change of the flex-direction seemed like a good approach to me

georgiee replied:
ok good I have no agenda to force order 🙏

georgiee commented on src/components/image-contentBlock/image-contentBlock.tsx:
```diff
@@ -0,0 +1,116 @@
+        {children}
+        {attribution && (
+          <AttributionContainer>
+            Photo by{' '}
```
This is not covered by the translation system. Have you though about it ?

milljoniaer replied:
I did not thought about it (probably since in Figma there is only "Photo by") and I thought that for this creator tag the english should then be fine (since the rest of the design is in german)...
Shall I support translation ? And who do I do this?

georgiee replied:
Don't trust the design too much 🤓 Think about the final result and imaging being a Germany visitor. Would you like to read random english words while the page is German? Please fix.

Please ask @hofmann-felix for support as he used it before too. We use https://react.i18next.com/ so please read the documentation in addition a little bit to know the foundation as this will hunt you in many places anyway 👹

milljoniaer replied:
Okay, yes, I implemented localisation

georgiee replied:
thanks 🙌

georgiee commented on src/components/image-contentBlock/image-contentBlock.tsx:
```diff
@@ -0,0 +1,116 @@
+  return (
+    <BlockWrapper textAlign={textAlign}>
+      <ImageWrapper>
+        {children}
```
I think it's good to pass in the children. That way image & GatsbyImage works in the best case. I guess we will hit a few issues with the layout though as use superimpose a specific layout where the gatsby image might have different ways of including an image (block vs inline an such)

milljoniaer replied:
So, do we leave it like that?

georgiee replied:
yeah I like to give positive head-ups instead of criticizing only in PRs 👍

georgiee reviewed (changes requested):
Thanks for this.
1. I would like to ask you to refactor this specific component again. See my comment in the `TextWrapper`. You made it work 💪🌟 Now I would like to guide you into a way that is more modern by relying on flex or grid css. Ask the team in the chat for details if you need help.  This is the reason to reject this PR. I'm torn though. We want to move fast to get things done, so I can perfectly imagine to merge this PR with a **mandatory** refactoring directly afterwards. Let's talk what's the best.

2. I have also commented on the name. Something is off.

3. The component still feels weird with the text included. I would like to hear about the opinion of the team @satellytes/website-crew. Does it feel good to have a single image component in the storybook that ships with the layout + text possibility? I would only see the image + attribution as the core element. Everything besides that is an additional component. I encouraged to move forward with the current version to get things done, so I think it's okay but a discussion about this feels never wrong.

milljoniaer committed:
chore: kebap-case

milljoniaer commented:
@feedm3 
> Not sure if it's possible, but can you add a Storybook example with `GatsybImage`?

I tried it, but when I use GatsbyImage, the story disappears....

milljoniaer committed:
chore: implement requested changes

georgiee commented on src/components/image-content-block/image-content-block.stories.tsx:
```diff
@@ -0,0 +1,37 @@
+  description:
+    'Integer posuere erat a ante venenatis dapibus posuere velit aliquet. Morbi leo risus, porta ac consectetur ac, vestibulum at eros. ',
+  attribution: {
+    source: 'www.somelink.de',
```
always use `www.example.com` for random urls you use for testing. That's the official test domain by IANA. Nobody can ever register it, so it's safe. Imagine www.somelink.de points to some NSFW content in the future, I think you don't want to be the person to have put it here 🤓

The same is valid and even more important fore mails. Don't use random mail addresses when testing or even sending mails. Always use `info@example.com` and such. The domain `example.com` again is the important part. Otherwise you annoy other people or leak internal information from some client.

milljoniaer replied:
That is a point, thank you for the information about ww.example.com 👍 
I changed it.

georgiee commented on src/components/image-content-block/image-content-block.stories.tsx:
```diff
@@ -0,0 +1,37 @@
+      control: { type: 'radio' },
+    },
+  },
+} as ComponentMeta<typeof Image>;
```
this part is usually ignore if you provide a mdx files so you can omit it. You define this through the meta tag in the mdx file then.

milljoniaer replied:
Oh, yes, that was a fail, fixed it 👍

georgiee commented on src/components/image-content-block/image-content-block.tsx:
```diff
@@ -0,0 +1,116 @@
+  width: 100%;
+
+  ${up('sm')} {
+    flex-direction: ${({ textAlign }) => mapTextAlignToFlex(textAlign)};
```
much better with the external mapping and the flex layout as the base 👍

milljoniaer replied:
👍

milljoniaer committed:
chore: internalization support

milljoniaer committed:
Merge branch 'main' into feat/image-contentblock

milljoniaer committed:
Merge branch 'main' into feat/image-contentblock

georgiee commented on src/components/image-content-block/image-content-block.tsx:
```diff
@@ -0,0 +1,117 @@
+      textAlign !== 'bottom' &&
+      css`
+        ${TextStyles.textS}
+        max-width: 228px;
```
we have a lack of requirements. I would have defined this as a shrinking and growing flex area with a percentage. But actually it feels okay with the fixed size so let's keep it.

milljoniaer replied:
I have already talked to Fabian about that, and he also said, that it is okay for now 👍

georgiee commented on src/components/image-content-block/image-content-block.tsx:
```diff
@@ -0,0 +1,117 @@
+  if (textAlign === 'right') {
+    return 'row';
+  } else if (textAlign === 'left') {
+    return 'row-reverse';
```
I mentioned the `order` attribute which is much more explicit than just reversing everything. This works here as we only have two elements but remember the `order` property to swap children to arbitrary places in the flex layout 👍

milljoniaer replied:
Okay, yes, this way felt quite cool to me, since I could also do the bottom thing within.
But I will read about `order` for the next time.

georgiee commented on src/components/image-content-block/image-content-block.tsx:
```diff
@@ -0,0 +1,117 @@
+}
+
+const BlockWrapper = styled.div<TextAlignProps>`
+  display: inline-flex;
```
Why would you define this an an inline container? Just make it flex which renders as a block to spread to the full width. That way you can also remove the width 100%.

milljoniaer replied:
Way cleaner, thank you 👍

georgiee commented on src/components/image-content-block/image-content-block.tsx:
```diff
@@ -0,0 +1,117 @@
+    ${({ textAlign }) =>
+      textAlign === 'bottom' &&
+      css`
+        ${TextStyles.textXS}
```
Interesting that the text has different sizes even though it's displayed for the same image in the same viewport. That just makes no sense to me and I have expected a unique size across all variants.

milljoniaer replied:
Yes it seems a little bit weird (but it is just like that Figma).

georgiee commented on src/components/image-content-block/image-content-block.tsx:
```diff
@@ -0,0 +1,117 @@
+
+interface ImageContentBlockProps {
+  children: React.ReactNode;
+  textAlign: 'right' | 'left' | 'bottom';
```
I really don't like the name `textAlign` because you are changing the layout and not the text alignment. You are creating a false expectation from the API. I would name this `textPlacement` or maybe you have some other alternative.

milljoniaer replied:
👍

georgiee commented on src/locales/de/translations.json:
```diff
@@ -160,5 +160,8 @@
       "link": "Karriere"
     }
   },
+  "image": {
+    "attribution": "Foto von "
+  },
```
Ideally you should provide an interpolation for things like that as you can't know the sentence structure of the future translation. See https://www.i18next.com/translation-function/interpolation. Something like `"key": "{{what}} is {{how}}`.

For here it's okay and makes things easier as we have a component involved anyway.

milljoniaer replied:
I will remember that for next time 👍

georgiee commented on src/components/image-content-block/image-content-block.stories.mdx:
```diff
@@ -0,0 +1,23 @@
+# Image
+
+The ContentBlock/Image Component can be used to display pictures with a description and an optional `attribution` tag in the bottom right corner.
+The image itself is passed as children. (e.g. as GatsbyImage Component, but always with `width: 100%;`for correct layouting).
```
> but always with `width: 100%`

Is this something we expect or something we apply? It's not clear from the text.

milljoniaer replied:
We expect it, I have changed it

georgiee reviewed (changes requested):
Thanks for the next iteration. I still have a few remarks but it looks good already. Coudl you update this PR another time? Thank you!

milljoniaer committed:
Merge branch 'main' into feat/image-contentblock

milljoniaer committed:
chore: small code cleanup things

milljoniaer committed:
Merge branch 'main' into feat/image-contentblock

milljoniaer committed:
chore: move component to new components
rename files to `image`

milljoniaer commented:
Thank you for our feedback. I think everything should be implemented now.
Additionally I have changed the file names of the now called "image" component into "image.*" and moved the whole folder to the new components.

georgiee reviewed (approved):
Let's merge that one finally. Thanks for your efforts

fischeversenker commented on src/new-components/image/image.stories.tsx:
```diff
@@ -0,0 +1,26 @@
+    source: 'https://www.example.com',
+    creator: 'Banksy',
+  },
+};
```
Let's only use the `mdx` format for documentation. You can define this story in the `stories.mdx` file so we can completely remove the `stories.tsx` file.

georgiee replied:
oh snap, you are right I missed this

fischeversenker commented on src/new-components/image/image.tsx:
```diff
@@ -0,0 +1,116 @@
+`;
+
+interface ImageContentBlockProps {
+  children: React.ReactNode;
+  textAlign: 'right' | 'left' | 'bottom';
+  description: string;
+  attribution?: {
+    source: string;
+    creator: string;
+  };
```
I suggest to do this the other way around: the image should be passed as a property, the description should be the children. Why? Do we know if it's always enough to have `description` as a plain string? Do we want to be able to have bold/italic text in the Image description? Also, it is really the "description" of the image or is it just some content that happens to be accompanied by an image?
I do understand that having the "description" passed in as a property makes it a lot easier to deal with the different font-sizes and responsiveness... but still, this feels upside down.

I'm sorry for the late but fundamental feedback, but I'd really like to see this topic addressed. Or you convince me that this way (Image as child, description as property) is solid?

georgiee replied:
I think we are heading in some wrong direction.
1. I don't want to use an image tag like this: `<Image image={}>lorem ipsim</Image>`
2. I never wanted to have an image with text block combine with attribution but I forced myself to sit tight until we have more time for refactorings.

Maybe we should go back two steps, what do we want to solve?

1. Attribution. Something I desperately wanted in the blog so we have it here. Maybe a simple attribution wrapper helps here or we create a [higher order component (hoc)](https://reactjs.org/docs/higher-order-components.html) for this specific purpose?
2. Text. This is probably what Felix writes about. From user perspective I see this:

```
<ImageWithCaption image={}>
 lorem ipsum
</ImageWithCaption>
```

This could also work as a HOC too wrapping any image.


This means if we rename this component to `ImageWithCaption` and swap the props description & children(image) we should be fine. In addition we should remove the attribution from this component and let some other component handle it. Best case:

```
const ImageWithCaptionAndAttribution = withAttribution(ImageWithCaption)

<ImageWithCaptionAndAttribution image={} attribution={}>
 lorem ipsum
</ImageWithCaptionAndAttribution>
```

that's a nice DX and a great split of responsibilties. What do you think?

fischeversenker replied:
Where else would you use that `withAttribution` hoc? The concept is nice... but in this particular case it looks very much overengineered to me.
I do like the very explicit `<ImageWithCaption/>`. However, I think whether the `<Image/>` that this component receives contains an attribution or not should not bother the `ImageWithCaption` component. I'd rather have an `ImageWithAttribution` component that can be used as a regular `Image` everywhere. Whether this is `withAttribution(Image)` or just a new regular component doesn't matter. I'd be careful with these advanced concepts like HOCs for our rather simple website. I'd very much argue that we want our website to be as junior-friendly as possible.

fischeversenker reviewed (commented):
Well. Now it's already merged :joy:

I admit that this feedback might come too late. I would have loved to see a few things addressed in here. Maybe these are now topics for a future refactoring of this new component :sparkles:
